azpubsub/contextvalidator/src/main/java/org/apache/kafka/common/security/authenticator/AzPubSubPrincipalBuilder.java

```diff
@@ -0,0 +1,136 @@
+
+            if(null != sslAuthenticationContextValidatorClass) {
+                LOGGER.info("Ssl Authentication Context Validator class is {}", KafkaConfig.AzpubsubSslAuthenticationValidatorClassProp());
+
+                sslAuthenticationContextValidator = (SslAuthenticationContextValidator) Utils.newInstance(sslAuthenticationContextValidatorClass);
+
+                if (null != sslAuthenticationContextValidator) {
+                    sslAuthenticationContextValidator.configure(configs);
+                }
+                else {
+                    LOGGER.error("Class {} provider by {} cannot be found or initialized. {}", sslAuthenticationContextValidatorClass, KafkaConfig.AzpubsubSslAuthenticationValidatorClassDoc(), KafkaConfig.AzpubsubSslAuthenticationValidatorClassDoc());
+
+                    throw new IllegalArgumentException("Class " + sslAuthenticationContextValidatorClass + " provided by setting " + KafkaConfig.AzpubsubSslAuthenticationValidatorClassProp()  + " is not found or cannot be initialized. " + KafkaConfig.AzpubsubSslAuthenticationValidatorClassDoc());
+                }
+            }
+
+            saslAuthenticationContextValidatorClass = Class.forName(configs.get(KafkaConfig.AzPubSubSaslAuthenticationValidatorClassProp()).toString());
+
+            if (null != saslAuthenticationContextValidatorClass) {
+
+                saslAuthenticationContextValidator = (SaslAuthenticationContextValidator) Utils.newInstance(saslAuthenticationContextValidatorClass);
+
+                if(null != saslAuthenticationContextValidator) {
+                    saslAuthenticationContextValidator.configure(configs);
+                }
+                else {
+                    LOGGER.error("Class {} provided by setting {} cannot be found or initialized. {}", saslAuthenticationContextValidatorClass, KafkaConfig.AzPubSubSaslAuthenticationValidatorClassProp(), KafkaConfig.AzPubSubSaslAuthenticationValidatorClassDoc());
+
+                    throw new IllegalArgumentException("Class " + saslAuthenticationContextValidatorClass + " provided by setting " + KafkaConfig.AzPubSubSaslAuthenticationValidatorClassProp() + " is not found. " + KafkaConfig.AzPubSubSaslAuthenticationValidatorClassDoc());
+                }
+            }
+        }
+        catch(ClassNotFoundException ex) {
+           throw new IllegalArgumentException(ex.getClass() + ". Error Message: " + ex.getMessage());
+        }
+    }
+
+    public KafkaPrincipal build(AuthenticationContext context) {
+
+        if ( context instanceof PlaintextAuthenticationContext ) {
+
+            return KafkaPrincipal.ANONYMOUS;
+        }
+        else if ( context instanceof SslAuthenticationContext ) {
+
+            SslAuthenticationContext sslAuthenticationContext = (SslAuthenticationContext)context;
+
+            if ( null != sslAuthenticationContextValidator ) {
```

Comment:
These nested if else should be documented. It is very confusing, even if you understand it now, after a few months I am sure you will forget. The other thing here is, did we cover all scenarios. I suggest to add a truth table here to document all possible parameter combinations.

Reply:
Also, what type of exception is thrown should be justified.

Reply:
The logic is quite simple here: call the validating method and return AzPubSubPrincipal. I did have some unit test cases there mocking the scenarios.